tee-supplicant: send READY=1 notification to systemd

embetrix opened the pull request:
This is very useful when tee-supplicant is started from systemd and can used with **Type=notify** to signal readiness

https://github.com/OP-TEE/optee_client/issues/382

jforissier commented on flags.mk:
```diff
@@ -9,6 +9,8 @@ PKG_CONFIG	?= $(CROSS_COMPILE)pkg-config
 
 C_COMPILER=$(shell readlink -f $$(which $(CC)))
 
+override LDFLAGS += -ldl
```
`dl` needs to be added to `tee-supplicant/CMakeLists.txt`, too (it is already there but conditional)

jforissier commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -493,6 +494,8 @@ static int usage(int status)
 	fprintf(stderr, "\t-d, --daemonize: run as a daemon (fork and return "
 			"after child has opened the TEE device or on error)\n");
+	fprintf(stderr, "\t-n, --sdnotify: signal READY=1 to systemd "
+			"after tee-supplicant is setup)\n");
```
Is this option really needed? What if we notify always?

jenswi-linaro replied:
Good question.

poettering replied:
my recommendation (as systemd upstream): just always send the notification, it internally conditionalizes things anyway based on an env var that is either passed or not, hence a second level of conditionalization should really not be necessary

jforissier commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -923,6 +931,28 @@ int main(int argc, char *argv[])
 		}
 	}
 
+	if ((sdnotify) && (!daemonize)) {
```
What prevents from notifying when `deamonize` is true?

jenswi-linaro replied:
The extra `()` are not needed.

jenswi-linaro commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -835,6 +839,7 @@ int main(int argc, char *argv[])
 		/* long name      | has argument  | flag | short value */
 		{ "help",            no_argument,       0, 'h' },
 		{ "daemonize",       no_argument,       0, 'd' },
+		{ "sdnotify",        no_argument,       0, 'n' },
```
`-n` usually means dry-run in other utilities. How about `-s` instead?

jenswi-linaro commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -923,6 +931,28 @@ int main(int argc, char *argv[])
 
+	if ((sdnotify) && (!daemonize)) {
+		/* we are set here notify systemd */
+		int(*__sd_notify__)();
```
The `__` are ugly, I suppose you've added them to avoid conflicts with `sd_notify()` in `<systemd/sd-daemon.h>`. How about `my_sd_notify(int unset_environment, const char *state)`?
Please initialize all local variables.

jenswi-linaro commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -923,6 +931,28 @@ int main(int argc, char *argv[])
+		int(*__sd_notify__)();
+		void *systemd = dlopen("libsystemd.so", RTLD_LAZY);
+		if (systemd) {
+			*(int**)(&__sd_notify__) = dlsym(systemd, "sd_notify");
```
Why do you need all this casting?

jenswi-linaro commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -923,6 +931,28 @@ int main(int argc, char *argv[])
+			*(int**)(&__sd_notify__) = dlsym(systemd, "sd_notify");
+			if (__sd_notify__) {
+					int ret = __sd_notify__(0, "READY=1");
+					if (ret <= 0) {
```
The extra `{}` are not needed.

embetrix commented:
@jforissier : I just figured out on the target testing that yocto doesn't ship libsystemd.so

but only versioned *so.*

```
# ls -l /lib/libsystemd.so*
lrwxrwxrwx 1 root root     20 Jul  9  2020 /lib/libsystemd.so.0 -> libsystemd.so.0.33.0
-rwxr-xr-x 1 root root 989248 Jul  9  2020 /lib/libsystemd.so.0.33.0
```

jforissier commented:
@embetrix `dlopen("libsystemd.so.0", RTLD_LAZY);` should work fine. Using `libsystemd.so` is actually not a good idea because you don't know which major version it is. It could lead to compatibility issues.

jenswi-linaro commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -923,6 +931,27 @@ int main(int argc, char *argv[])
+					if (ret <= 0)
+						fprintf(stderr, "sd_notify failed: %d\n", ret);
+			}
+			else
```
Please fold up `else` to make it `} else`

embetrix commented:
> @embetrix `dlopen("libsystemd.so.0", RTLD_LAZY);` should work fine. Using `libsystemd.so` is actually not a good idea because you don't know which major version it is. It could lead to compatibility issues.

done

jenswi-linaro commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -923,6 +931,27 @@ int main(int argc, char *argv[])
+		int ret = 0;
+		void *sd_handle = NULL;
+ 		int (*sd_notify)(int unset_environment, const char *state) =  NULL;
+		sd_handle = dlopen("libsystemd.so.0", RTLD_LAZY);
```
Please leave an empty line between variables and code.

jenswi-linaro commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -923,6 +931,27 @@ int main(int argc, char *argv[])
+	if (sdnotify && !daemonize) {
+		int ret = 0;
+		void *sd_handle = NULL;
+ 		int (*sd_notify)(int unset_environment, const char *state) =  NULL;
```
This is bound to cause a compile error if `<systemd/sd-daemon.h>` is ever included, that's why I suggested `my_sd_notify` or such.

mikkorapeli-linaro commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -923,6 +926,20 @@ int main(int argc, char *argv[])
+	if (sd_handle) {
+		tee_sd_notify = dlsym(sd_handle, "sd_notify");
+		if (tee_sd_notify) {
+				e = tee_sd_notify(0, "READY=1");
```
While at it, could you add a "tee_sd_notify(0, "WATCHDOG=1"); into the main loops which call process_one_request()?

What could be the maximum time that a process_one_request() call takes? We could then double this and add to systemd service file to detect if tee-supplicant hangs. I've seen hangs related ftpm kernel module unloading which could affect tee-supplicant too, for example.

jenswi-linaro replied:
How long does it take to copy a file (of unknown size) from the file system into memory, worst case?

mikkorapeli-linaro replied:
Ok, the watchdog target is application dependent but the sd_notify() call to kick the watchdog could be generic in the loop.

Users could set the application specific watchdog timeout in the systemd service file for tee-supplicant. Default could be that it's not set at all.

jforissier replied:
The `ioctl()` call is blocking until a request comes, with no timeout. So we would at least need to add a timeout before adding support for such a watchdog. An alternative would be to ping from a separate thread but that's obviously a weaker health guarantee.

mikkorapeli-linaro replied:
The watchdog call from tee-supplicant to systemd is just a health signal. It can block for ever. The timeout value is set in systemd service file, which can be unset by default, or a really large value. Dending on use cases and TAs, then this value can be reduced.

But it's ok to skip this for now too.

mikkorapeli-linaro commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -923,6 +926,20 @@ int main(int argc, char *argv[])
+					fprintf(stderr, "sd_notify failed: %d\n", e);
+		} else
+			fprintf(stderr, "Couldn't find sd_notify symbol: %s\n", dlerror());
+		dlclose(sd_handle);
```
If watchdog calls are done, I'd leave this open until exiting mainloops

poettering replied:
also note that libsystemd.so is nowadays built with `-z nodelete`, i.e. we do not support being unloaded, and the dlclose() is hence without effect

poettering commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -923,6 +926,20 @@ int main(int argc, char *argv[])
 	}
 
+	/* we are set here notify systemd */
+	sd_handle = dlopen("libsystemd.so.0", RTLD_LAZY);
```
sorry for parachuting in here, I was pointed to this PR from here: 

https://lists.freedesktop.org/archives/systemd-devel/2024-June/050385.html

I just wanted to mention that for a trivial call like sd_notify() you might just opt directly implementing this, rather than doing dlopen()/dlsym(). i.e. something like this:

https://www.freedesktop.org/software/systemd/man/devel/sd_notify.html#Standalone%20Implementations

poettering commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -923,6 +926,20 @@ int main(int argc, char *argv[])
+		if (tee_sd_notify) {
+				e = tee_sd_notify(0, "READY=1");
+				if (e <= 0)
+					fprintf(stderr, "sd_notify failed: %d\n", e);
```
please do not consider `e == 0` an error, it's really not. It just means no message sent because $NOTIFY_SOCKET wasn't set, i.e. because it was invoked in a non-systemd context, which is entirely fine however.

jforissier commented on tee-supplicant/CMakeLists.txt:
```diff
@@ -28,6 +28,7 @@ set(SRC
 	src/hmac_sha2.c
 	src/rpmb.c
 	src/sha2.c
+	src/tee_sd_notify.c
```
Let's drop the `tee_` prefix, it's distraction

jforissier commented on tee-supplicant/src/tee_sd_notify.c:
```diff
@@ -0,0 +1,81 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
```
```C
// SPDX-License-Identifier: MIT-0
/*
 * The code below is imported from:
 * https://www.freedesktop.org/software/systemd/man/devel/sd_notify.html#Standalone%20Implementations
 */

...
```

jforissier commented on tee-supplicant/src/tee_sd_notify.h:
```diff
@@ -0,0 +1,28 @@
+#ifndef TEE_SD_NOTIFY_H
+#define TEE_SD_NOTIFY_H
+
+#define _GNU_SOURCE 1
+#include <string.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <signal.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <time.h>
+#include <unistd.h>
```
Move to `sd_notify.c`

jforissier commented on tee-supplicant/src/tee_sd_notify.c:
```diff
@@ -0,0 +1,81 @@
+    return 1; /* Notified! */
+}
+
+int tee_sd_notify_ready(void)
```
`sd_notify_ready()`

mikkorapeli-linaro commented:
Tested latest version of patch with TS and TRS in https://ledge.validation.linaro.org/scheduler/job/88692 and it works. tee-supplicant startup state is correctly reported to systemd.

jforissier commented on tee-supplicant/src/sd_notify.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef SD_NOTIFY_H
+#define SD_NOTIFY_H
+
+#define _GNU_SOURCE 1
+#include <string.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <signal.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <time.h>
+#include <unistd.h>
```
Move the includes to `sd_notify.c`

jforissier commented on tee-supplicant/src/tee_supplicant.c:
```diff
@@ -923,6 +924,12 @@ int main(int argc, char *argv[])
+	e = sd_notify_ready();
+	if(e < 0) {
+		fprintf(stderr, "sd_notify_ready failed: %s\n", strerror(-e));
+	}
```
Style:
- `{ }` not needed.
- "sd_notify_ready() failed" (parentheses after function names)

jforissier reviewed (commented):
- See comments below.
- Please use prefix in commit subject: "tee-supplicant: send READY=1 notification to systemd"
- Please add your `Signed-off-by:`

embetrix commented:
> * See comments below.
> * Please use prefix in commit subject: "tee-supplicant: send READY=1 notification to systemd"
fixed
> * Please add your `:`
added Signed-off-by

jforissier reviewed (approved):
Please wrap the commit description at 72 characters and add:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

@poettering thanks for your recommendations

embetrix commented:
> Please wrap the commit description at 72 characters and add:
> 
> `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`
> 
> @poettering thanks for your recommendations

done

jforissier commented:
Any more feedback? @mikkorapeli-linaro would you mind giving an Ack and/or Tested-by perhaps?

mikkorapeli-linaro commented:
> Any more feedback? @mikkorapeli-linaro would you mind giving an Ack and/or Tested-by perhaps?

Yes, latest patch version also

Tested-by: Mikko Rapeli <mikko.rapeli@linaro.org>

TS firmware with optee 4.1 and RPMB kernel patches from Jens, and TRS uki kernel and initramfs with systemd which starts tee-supplicant using this patch to create fTPM encrypted rootfs, on rockpi4b. The few remaining failures in TS/TRS test suite are not related to this patch.

Test run:
https://ledge.validation.linaro.org/scheduler/job/88800

jforissier commented:
Mikko's `Tested-by:` tag added and pushed to master. Thanks!
